test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs

```diff
@@ -0,0 +1,211 @@
+        /// </summary>
+        /// <param name="span">The span on which to filter.</param>
+        /// <returns>Whether the span qualifies for this expectation.</returns>
+        public virtual bool ShouldInspect(MockTracerAgent.Span span)
+        {
+            return span.Service == ServiceName
+                && span.Name == OperationName
+                && span.Type == Type;
+        }
+
+        /// <summary>
+        /// The aggregate assertion which is run for a test.
+        /// </summary>
+        /// <param name="span">The span being asserted against.</param>
+        /// <param name="message">The developer friendly message for the test failure.</param>
+        /// <returns>Whether the span meets expectations.</returns>
+        public bool MeetsExpectations(MockTracerAgent.Span span, out string message)
+        {
+            message = string.Empty;
+
+            var messages = new List<string>();
+
+            foreach (var assertion in Assertions)
+            {
+                var mismatchMessage = assertion(span);
+                if (!string.IsNullOrWhiteSpace(mismatchMessage))
+                {
+                    messages.Add(mismatchMessage);
+                }
+            }
+
+            if (messages.Any())
+            {
+                message = string.Join(",", messages);
+                return false;
+            }
+
+            return true;
+        }
+
+        public void TagShouldExist(string tagKey, Func<MockTracerAgent.Span, bool> when)
+        {
+            Assertions.Add(span =>
+            {
+                if (when(span) && !span.Tags.ContainsKey(tagKey))
+                {
+                    return $"Tag {tagKey} is missing from span.";
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterDelegateExpectation(Func<MockTracerAgent.Span, IEnumerable<string>> expectation)
+        {
+            Assertions.Add(span =>
+            {
+                var failures = expectation(span);
+
+                if (failures.Any())
+                {
+                    return string.Join(",", failures);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterCustomExpectation(string keyForMessage, Func<MockTracerAgent.Span, string> actual, Func<MockTracerAgent.Span, string> expected)
+        {
+            Assertions.Add(span =>
+            {
+                var actualValue = actual(span);
+                var expectedValue = expected(span);
+
+                if (expectedValue != null && actualValue != expectedValue)
+                {
+                    return FailureMessage(name: keyForMessage, actual: actualValue, expected: expectedValue);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterTagExpectation(string key, string expected, Func<MockTracerAgent.Span, bool> when, Func<string, string> customMessage = null)
```

Comment:
Could you rename `customMessage` to `customErrorMessageDelegate` or something comparable? That would help me with understanding the code more easily

Reply:
Also, is `customMessage` used anywhere? Or is that just added more future extensibility?

Reply:
It was used before, I can delete it.